### PR TITLE
fix(dynamics): correct MTK NPT/NPH barostat thermostat coupling and Q_b sizing

### LINKED ISSUE #89: 🐛[BUG]: NPT cell-volume drifts past ASE/TorchSim equilibrium density without converging
# 🐛 [BUG]: NPT cell-volume drifts past the ASE/TorchSim equilibrium density without converging

## Summary

In NPT simulations, the  `NPT` integrator does not equilibrate to the same density that ASE NPT and TorchSim NPT reach under identical model and target-state settings. The cell continues to expand and the density falls past the reference plateau without recovering. Behaviour is qualitatively the same across `pressure_coupling` modes (`isotropic`, `anisotropic`, `triclinic`); switching mode only changes the rate of drift.

The bug has been reproduced on two independent systems:

1. **Liquid n-octane + AIMNet2 (B97-3c) + DFTD3 + Ewald**, 20 molecules / 520 atoms, T=298 K, P=1 atm. Reference density (NIST WebBook): ~0.703 g/cm³. ASE NPT under the same calculator equilibrates around ~0.74 g/cm³. Toolkit NPT density falls from 0.84 → 0.18 g/cm³ over 35 000 steps with no plateau. Reproducer + dashboard image attached below.

2. **Naphthalene 2×2×2 supercell + AIMNet2 (no Ewald, no D3)**, 288 atoms, T=300 K, P=1 atm. Toolkit NPT density falls from 1.20 → 0.36 g/cm³ over 10 000 steps; ASE NPT under the same calculator at 5 000 steps is at 0.81 g/cm³ and crossing zero pressure.

## Environment

- Repo: `NVIDIA/nvalchemi-toolkit` at upstream `7fe7756 fix(models): merge force and stress autograd (#88)`
- Hardware: NVIDIA GH200 (CUDA 13)
- Stack: Python 3.12, PyTorch CUDA 13.0 wheel, `aimnet`, `ase`, `nvalchemiops` matching toolkit version

---

## System A — Liquid n-octane (full pipeline)

Pipeline: `PipelineModelWrapper(groups=[PipelineGroup([AIMNet2, EwaldModelWrapper], use_autograd=True), PipelineGroup([DFTD3ModelWrapper])])`. NPT params: `dt=0.5 fs`, `barostat_time=5000 fs`, `thermostat_time=100 fs`, `pressure_coupling="anisotropic"`, `chain_length=3` (default).

Selected rows from `npt_prod.csv` (full file in attachment):

```
step    volume(Å³)   density(g/cm³)  pressure_atm   a(Å)     alpha   beta   gamma
   0    39567.67     0.5129          -4075.29       34.08    90.00   90.00  90.00
 100    39570.41     0.5129          -2860.94       34.08    90.00   90.00  90.00
 ...
13700   54808.79     0.3703          -1487.02       37.99    90.00   90.00  90.00
13900   54993.43     0.3691          -1420.65       38.03    90.00   90.00  90.00
```

Notes:
- Cell angles stay at 90° throughout — cell remains orthorhombic, so this is not a non-axis-aligned-cell artefact.
- `pressure_atm` is computed (in `main.py:_compute_pressure_eV_A3`) as `Tr(batch.stress)/3 + (2/3)·KE/V`, treated as compression-positive per the script comment. The reported value is consistently negative throughout the run (ranging −4 000 … −1 500 atm).
- Cell volume grows monotonically (`(e) Cell volume` in dashboard); density falls monotonically (`(f) Density`) and is heading well below ASE's ~0.74 plateau by the end of the run.
- Temperature is held at the target by the thermostat (`(b)`), forces and energy stay finite (`(a),(d)`), so the run is not blowing up numerically — it is sampling at progressively lower densities.

---

## System B — Naphthalene crystal (bare AIMNet2)

Self-contained reproducer (~80 lines, no Ewald/D3, single graph) using the experimental P21/a unit cell from `alchemi-playbooks/dev/part-2-toolkit/data/naphthalene.cif`. 2×2×2 supercell, 288 atoms.

| Integrator / mode | barostat_time | thermostat_time | V(0) Å³ | V(5 000) Å³ | density(0) | density(5 000) |
|---|---|---|---|---|---|---|
| Toolkit NPT, `pressure_coupling="anisotropic"` | 2 000 fs | 100 fs | 2 842 | 4 015 (+41 %) | 1.198 | 0.848 |
| Toolkit NPT, `pressure_coupling="isotropic"`   | 2 000 fs | 100 fs | 2 842 | 3 477 (+22 %) | 1.198 | 0.980 |
| Toolkit NPT, `pressure_coupling="triclinic"`   | 2 000 fs | 100 fs | 2 842 | 4 993 (+76 %) | 1.198 | 0.682 |
| Toolkit NPT, `pressure_coupling="triclinic"`   | 2 000 fs | 100 fs | 2 842 | 9 399 (+231 %) at 10 000 steps | 1.198 | 0.362 (still falling) |
| ASE `MelchionnaNPT` (reference)                | 2 000 fs | 100 fs | 2 842 | 4 219 (+48 %) | 1.198 | 0.807 |

For `triclinic` mode at user-realistic `barostat_time=2 000 fs`, the trace `Tr(stress)/3` swings from +0.024 eV/Å³ (compressive) at step 200 down through zero at step ~2 200, becomes negative (tensile) around steps 2 200–4 000, then stays mildly negative for the rest of the 10 000 steps even as the cell continues to expand. ASE under the same calculator reaches its reference plateau (~0.81 → ~0.74 g/cm³) within 5 000–20 000 steps and stops there.

### Stress sign sanity check

For the same initial geometry, lab-frame diagonal Cauchy stress components from `batch.stress` (toolkit, bare AIMNet2) vs. `atoms.get_stress(voigt=True)` (ASE+AIMNet2) are sign-flipped, as expected from each library's documented convention:

```
xx: toolkit=+4.575e-02  ASE=-3.033e-02   OPPOSITE-SIGN (OK)
yy: toolkit=+1.996e-01  ASE=-1.777e-01   OPPOSITE-SIGN (OK)
zz: toolkit=+1.522e-01  ASE=-1.476e-01   OPPOSITE-SIGN (OK)
```

So the bare-AIMNet2 stress on `batch.stress` matches expectations (compression-positive Cauchy, sign-flipped relative to ASE's tension-positive).

---

## Reproducer (System B, self-contained)

```python
# diag_step3.py
"""NPT volume-drift reproducer.
Usage:
  python diag_step3.py {anisotropic|isotropic|triclinic} <n_steps> <log_every>
"""
from __future__ import annotations
import sys
import torch
from ase.io import read as ase_read

from nvalchemi.data import AtomicData, Batch
from nvalchemi.dynamics import initialize_velocities
from nvalchemi.dynamics.integrators.npt import NPT
from nvalchemi.models.aimnet2 import AIMNet2Wrapper

torch.manual_seed(42)
unit_cell = ase_read("naphthalene.cif")  # P21/a, β≈124°, from alchemi-playbook
supercell = unit_cell * (2, 2, 2)
n = len(supercell)

device = "cuda"
data = AtomicData.from_atoms(supercell)
data.forces = torch.zeros(n, 3)
data.energy = torch.zeros(1, 1)
data.stress = torch.zeros(1, 3, 3)
data.add_node_property("velocities", torch.zeros(n, 3))
data.charge = torch.zeros(1, 1)
batch = Batch.from_data_list([data], device=device)

aimnet2 = AIMNet2Wrapper.from_checkpoint("aimnet2", device=device, compile_model=False)
aimnet2.set_config("active_outputs", {"energy", "forces", "stress"})

T_K = 300.0
T_per_sys = torch.full((1,), T_K, device=device, dtype=batch.positions.dtype)
initialize_velocities(
    batch.velocities, batch.atomic_masses, T_per_sys, batch.batch_idx.int(), random_seed=42
)

P_1ATM = 101325.0 / 1.602176634e11  # eV/Å³
mode = sys.argv[1] if len(sys.argv) > 1 else "anisotropic"
N_STEPS = int(sys.argv[2]) if len(sys.argv) > 2 else 5000
LOG_EVERY = int(sys.argv[3]) if len(sys.argv) > 3 else 200

if mode == "anisotropic":
    pressure = torch.tensor([[P_1ATM, P_1ATM, P_1ATM]], dtype=torch.float32)
elif mode == "isotropic":
    pressure = torch.tensor([P_1ATM], dtype=torch.float32)
elif mode == "triclinic":
    pressure = torch.eye(3, dtype=torch.float32).unsqueeze(0) * P_1ATM
else:
    raise SystemExit(mode)

npt = NPT(
    model=aimnet2,
    dt=0.5,
    temperature=T_K,
    pressure=pressure,
    barostat_time=2000.0,
    thermostat_time=100.0,
    pressure_coupling=mode,
    chain_length=3,
)
for h in aimnet2.make_neighbor_hooks():
    npt.register_hook(h)

total_mass = batch.atomic_masses.sum().item()
density_factor = total_mass * 1.660539e-24 * 1.0e24
def density_g_cc(v): return density_factor / v

print("step   V (Å³)    density (g/cc)   Tr(stress)/3 (eV/Å³)")
for step in range(N_STEPS + 1):
    if step % LOG_EVERY == 0:
        V = torch.linalg.det(batch.cell[0]).abs().item()
        Tr3 = batch.stress[0].cpu().diagonal().mean().item() if step > 0 else 0.0
        print(f"{step:5d}   {V:8.2f}   {density_g_cc(V):8.4f}   {Tr3:+.3e}")
    if step == N_STEPS:
        break
    batch, _ = npt.step(batch)
print("DONE.")
```

ASE reference:

```python
from ase.md.npt import NPT as ASENPT
from ase import units
import numpy as np
from ase.md.velocitydistribution import MaxwellBoltzmannDistribution

# atoms = supercell as above with AIMNet2ASE("aimnet2") calculator
MaxwellBoltzmannDistribution(atoms, temperature_K=300, rng=np.random.default_rng(42))
dyn = ASENPT(
    atoms,
    timestep=0.5 * units.fs,
    temperature_K=300,
    externalstress=-1.01325 * units.bar,        # 1 atm, ASE tension-positive convention
    ttime=100.0 * units.fs,
    pfactor=(2000.0 * units.fs) ** 2 * 10.0 * units.GPa,
    mask=np.diag([1, 1, 1]),
)
dyn.run(5000)
```

The naphthalene CIF (`a=8.21, b=5.97, c=8.67 Å, β=124.4°`, P21/a) is the experimental unit cell attached here.

### PR DESCRIPTION
**Pull request by Ryan-Reese**
## Description

Fixes four bugs in the toolkit's MTK NPT integrator (`nvalchemi/dynamics/integrators/npt.py`) that combine to drive unbounded cell-volume drift in long NPT runs (#89). The `W` normalization is also fixed in `nvalchemi/dynamics/integrators/nph.py`. All four are cross-validated against `ase.md.nose_hoover_chain` (`IsotropicMTKNPT`, `MTKNPT`) and `torch_sim.integrators.npt.npt_nose_hoover_isotropic`.

| bug | location | fix |
|---|---|---|
| Barostat mass `W` was only divided by 3 in non-isotropic modes; iso kept the full kernel-emitted `(3N+3)·kT·τ_P²` (3× too large vs canonical) | `npt.py:205, 268`, `nph.py:146, 180` | drop the conditional and always divide `W` by 3 |
| Barostat-thermostat NHC kernel triples its DOF input, so caller passes pre-divided `cell_ndof / 3` | `npt.py:387, 502` | divide `_cell_dof_count(...)` by 3 before the NHC call |
| Cell-velocity update applies a spurious particle-thermostat drag inside the barostat half-step | `npt.py:421, 489` | pass `torch.zeros_like(...)` for the `eta_dots` argument |
| `Q_b` (barostat NHC mass) was sized with `τ_T` (particle-thermostat time) instead of `τ_P` (barostat time) | `npt.py:222` | pass `tau_p` to `nhc_compute_masses` |

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Related Issues

Fixes #89.

NVIDIA/nvalchemi-toolkit-ops#77 lands the kernel-side fix that lets the `nhc_zero_eta_dot` workaround be removed; once that ships, a small follow-up PR will drop the buffer here.

## Changes Made

- `nvalchemi/dynamics/integrators/npt.py` — four edits matching the table above.
- `nvalchemi/dynamics/integrators/nph.py` — applies the same `W / 3` normalization in `_init_state` and `_make_new_state`.
- `_cell_kinetic_energy` — deprecate `cells_inv` (kept for backward compat, emits `DeprecationWarning`); helper reads `cell_velocity` directly as the strain rate.
- `test/dynamics/test_npt_nph.py` — replaces `test_W_divided_for_non_isotropic` with `test_W_normalization_matches_isotropic_across_modes` (asserts `iso == aniso == triclinic`); adds `test_Q_b_scales_with_barostat_time_not_thermostat_time`.
- `CHANGELOG.md` — `Fixed` and `Deprecated` entries.

## Testing

- [x] Unit tests pass locally (`make pytest`) — 38/38.
- [x] Linting passes (`make lint`).
- [x] Regression tests added: `test_W_normalization_matches_isotropic_across_modes` and `test_Q_b_scales_with_barostat_time_not_thermostat_time` fail against `origin/main` and pass on this branch.

### Empirical validation

NPT comparison on a 100-molecule n-octane box (~2600 atoms, Packmol seed = 42, initial ρ ≈ 0.703 g/cm³, Orb-v3-conservative-inf-omat + DFT-D3(BJ), T = 298 K, P = 1 atm, dt = 0.5 fs, τ_T = 100 fs, τ_P = 5000 fs, NHC chain length 3). Toolkit runs use `pressure_coupling="anisotropic"` with `pressure=[[P, P, P]]`; ASE uses `MTKNPT` (full triclinic, 9-DOF — same `target/Q` ratio as toolkit aniso, giving identical NHC barostat oscillation frequency).

density_comparison_3way_post_codex_fix

NIST experimental density at 298 K, 1 atm: 0.703 g/cm³.

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes (inline comments at each fixed call site explain the underlying bug + canonical reference)
- [ ] I have updated the documentation (if applicable)

### CODE REVIEW

File: nvalchemi/dynamics/integrators/npt.py [hunk 1]
```diff
@@ -391,7 +397,7 @@ def pre_update(self, batch: Batch) -> None:
             self._state.W,
             KE,
             self._state.num_atoms_per_system,
-            self._state.nhc_eta_dot,
+            torch.zeros_like(self._state.nhc_eta_dot),
             self._state.dt,
         )
```
[the remark this answers is not in the document]
Comment (Ryan-Reese) [reply nvalchemi/dynamics/integrators/npt.py:425]: @greptileai

**Review by zubatyuk — CHANGES_REQUESTED** [review]
Minor fixes.

**Comment by zubatyuk** [comment]
LGTM

File: test/dynamics/test_npt_nph.py [hunk 2]
```diff
@@ -505,8 +514,39 @@ def test_W_divided_for_non_isotropic(self, mode, pressure):
         )
         batch = _make_barostat_batch()
         npt_iso._init_state(batch)
-        npt_aniso._init_state(batch)
-        assert torch.allclose(npt_iso._state.W / 3, npt_aniso._state.W, atol=1e-6)
+        npt_other._init_state(batch)
+        assert torch.allclose(npt_other._state.W, npt_iso._state.W, atol=1e-6)
+
+    def test_Q_b_scales_with_barostat_time_not_thermostat_time(self):
+        """Barostat NHC chain mass ``Q_b`` must be sized with ``τ_P``, not ``τ_T``.
```
Comment (laserkelvin) [inline test/dynamics/test_npt_nph.py:521]: I would make this comment simpler - just refer to the GitHub issue instead

File: nvalchemi/dynamics/integrators/npt.py [hunk 3]
```diff
@@ -383,6 +407,11 @@ def pre_update(self, batch: Batch) -> None:
 
         P_inst = self._compute_P(batch, volumes)
         KE = self._compute_ke(batch)
+        # Workaround for nvalchemi-toolkit-ops <= 0.3.1: pass zero ``eta_dots``
```
Comment (laserkelvin) [inline nvalchemi/dynamics/integrators/npt.py:410]: @Ryan-Reese do you mind creating an issue to track this?

**Review by laserkelvin — APPROVED** [review]
LGTM, the comment was more optional than anything else

File: nvalchemi/dynamics/integrators/npt.py [hunk 4]
```diff
@@ -230,6 +238,18 @@ def _init_state(self, batch: Batch) -> None:
...
+                # ``-eta_dot * h_dot`` drag, but canonical MTK puts no drag
+                # in the barostat half-step (cell damping comes from
+                # ``b_scale`` below).  Pass a permanent zero buffer to
+                # mute the drag.
+                # TODO: drop ``nhc_zero_eta_dot`` and switch to the
+                # eta_dots-free API once nvalchemi-toolkit-ops > 0.3.1
+                # ships (kernel fix in NVIDIA/nvalchemi-toolkit-ops#77).
+                "nhc_zero_eta_dot": torch.zeros(
+                    M, self.chain_length, dtype=dtype, device=dev
+                ),
```
[the remark this answers is not in the document]
Comment (Ryan-Reese) [reply nvalchemi/dynamics/integrators/npt.py:252]: @greptileai don't think this can happen `mutates_args={"cell_velocity"}`
